controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -212,16 +212,21 @@ func (r *BareMetalHostReconciler) Reconcile(request ctrl.Request) (result ctrl.R
 		return ctrl.Result{}, errors.Wrap(err, "failed to create provisioner")
 	}
 
-	ready, err := prov.IsReady()
+	if host.Spec.ExternallyProvisioned && !host.HasBMCDetails() {
+		// In the case of Assisted Installer, we don't have Ironic or BMC info.
+		// so the IsReady() check will fail every time on the ironic endpoint check.
+		reqLogger.Info("externally provisioned host has no BMC details, assuming it is ready")
```

Comment:
The thing being checked by IsReady() is the provisioner, not the host. So rather than logging that we assume it is ready, we should say we're ignoring the provisioner. We could also just not log at all, in which case I would flip the logic in line 215 so we don't need an else clause for the if statement.

Reply:
Ok, from the current code I understand that there are certain conditions on the host (when someone else is managing and there are no BMC details) where it doesn't make sense to have an (Ironic) provisioner (and probably the same for the state machine).
Maybe at this point could be worth to have a slightly more dinamic provisioner factory and create an ad-hoc provisioner type for this case, instead of putting specific checks?

Reply:
Having a separate provisioner implementation is interesting. There's basically nothing we can do with these hosts, though, since they are not completely configured. So aside from the `IsReady` check, I don't think any other provisioner methods would ever be called. Would they?

Reply:
we do set the provisioner factory up in main.go rather than in the Reconcile, which would make this messy. but i could move the check into the IsReady()?

Reply:
What about adding a new intermediate layer directly in the Ironic factory `Ironic.New`? It could be a good point to decide which implementation should be created, and it will not impact severely the current code architecture.

Reply:
Maybe I was confused about the goal of this PR.

When the host does not have credentials, we cannot put it through the rest of the state machine, right? We can't register it with Ironic, because Ironic requires credentials and other identifying details. We also don't want to end up in a state like Ready where we think we can control the hosts power. And a host without credentials isn't in an "error" state that means we want to retry operations, because we won't see different results just by trying again with no new data.

We want these hosts to sit quietly in the Unmanaged state until someone gives them credentials. Is there a bug with the way externally provisioned but unmanaged hosts are being handled?

Reply:
I think the main problem is that they do not have Ironic yet (in addition to the missing BMC details), so also the provisioner readiness check fails.
As per the current state machine implementation a non-working Provisioner potentially is not a problem in such case, since the host keeps stalling into the `Unmanaged` state.

Reply:
OK, I have hopefully made this less hacky/more clear now.

Reply:
Hey @asalkeld I like the approach and it seems it could work fine, I think that if you just make the factory wrapper a decorator there will be no impact at all on the factory signature